cranelift-faerie/src/backend.rs

```diff
@@ -74,9 +79,164 @@ pub struct FaerieBackend {
+            -0x08,
+            // ISA-specific, column for the return address (may be a register, may not)
+            reg_mapper.return_address(),
+        );
+
+        cie.fde_address_encoding = gimli::DwEhPe(0x1b);
+
+        let mut encoder = CFIEncoder::new(&reg_mapper);
+
+        for inst in initial_state
+            .iter()
+            .flat_map(|change| encoder.translate(change))
+        {
+            cie.add_instruction(inst);
+        }
+
+        let cie_id = self.table.add_cie(cie);
+
+        (cie_id, encoder)
+    }
+
+    pub fn new() -> FrameSink {
+        FrameSink {
+            fn_names: vec![],
+            table: gimli::write::FrameTable::default(),
+        }
+    }
+
+    pub fn address_for(&mut self, name: &str) -> Address {
+        // adding a FrameDescriptionEntry for a function twice would be a bug,
+        // so we can confidently expect that `name` will not be provided more than once.
+        // So `name` is always new, meaning we can just add it and return its index
+        self.fn_names.push(name.to_string());
+        Address::Symbol {
+            symbol: self.fn_names.len() - 1,
+            addend: 0,
+        }
+    }
+
+    /// Add a FrameDescriptionEntry to the FrameTable we're constructing
+    ///
+    /// This will always use the default CIE (which was build with this `FrameSink`).
+    pub fn add_fde(&mut self, cie_id: gimli::write::CieId, fd_entry: FrameDescriptionEntry) {
+        self.table.add_fde(cie_id, fd_entry);
+    }
+}
+
+struct FaerieDebugSink<'a> {
+    pub data: &'a mut Vec<u8>,
+    pub functions: &'a [String],
+    pub artifact: &'a mut faerie::Artifact,
+}
+
+impl<'a> gimli::write::Writer for FaerieDebugSink<'a> {
+    type Endian = gimli::LittleEndian;
+
+    fn endian(&self) -> Self::Endian {
+        gimli::LittleEndian
+    }
+    fn len(&self) -> usize {
+        self.data.len()
+    }
+    fn write(&mut self, bytes: &[u8]) -> gimli::write::Result<()> {
+        self.data.extend_from_slice(bytes);
+        Ok(())
+    }
+
+    fn write_at(&mut self, offset: usize, bytes: &[u8]) -> gimli::write::Result<()> {
+        if offset + bytes.len() > self.data.len() {
+            return Err(gimli::write::Error::LengthOutOfBounds);
+        }
+        self.data[offset..][..bytes.len()].copy_from_slice(bytes);
+        Ok(())
+    }
+
+    fn write_eh_pointer(
+        &mut self,
+        address: Address,
+        eh_pe: gimli::DwEhPe,
+        size: u8,
+    ) -> gimli::write::Result<()> {
+        // we only support PC-relative 4byte signed offsets for eh_frame pointers currently. Other
+        // encodings may be permissible, but aren't seen even by gcc/clang/etc, and have not been
+        // tested. Currently, relocations used for addresses expect to be relocating four bytes,
+        // PC-relative, and larger pointer sizes would require selection of other relocation types.
+        assert!(eh_pe.0 == 0x1b);
```

Comment:
Can you compare `eh_pe` against `DW_EH_PE_sdata4` instead?

Reply:
this'd be `DW_EH_PE_pcrel | DW_EH_PE_sdata4`, but yeah definitely